controllers/controllers/cluster_controller.go

```diff
@@ -32,32 +38,70 @@ const (
 
 // ClusterReconciler reconciles a Cluster object
 type ClusterReconciler struct {
-	client    client.Client
-	log       logr.Logger
-	validator *vsphere.Validator
-	defaulter *vsphere.Defaulter
-	tracker   *remote.ClusterCacheTracker
+	client                  client.Client
+	log                     logr.Logger
+	validator               *vsphere.Validator
+	defaulter               *vsphere.Defaulter
+	tracker                 *remote.ClusterCacheTracker
+	buildProviderReconciler BuildProviderReconciler
 }
 
-func NewClusterReconciler(client client.Client, log logr.Logger, scheme *runtime.Scheme, govc *executables.Govc, tracker *remote.ClusterCacheTracker) *ClusterReconciler {
+// TODO: this is not ideal and will need a refactor. I will follow up but for now this
+// allows us to decouple the cluster reconciler main logic from provider specific logic
+type BuildProviderReconciler func(datacenterKind string, client client.Client, log logr.Logger, validator *vsphere.Validator, defaulter *vsphere.Defaulter, tracker *remote.ClusterCacheTracker) (clusters.ProviderClusterReconciler, error)
+
+func NewClusterReconciler(client client.Client, log logr.Logger, scheme *runtime.Scheme, govc *executables.Govc, tracker *remote.ClusterCacheTracker, buildProviderReconciler BuildProviderReconciler) *ClusterReconciler {
 	validator := vsphere.NewValidator(govc, &networkutils.DefaultNetClient{})
 	defaulter := vsphere.NewDefaulter(govc)
 
 	return &ClusterReconciler{
-		client:    client,
-		log:       log,
-		validator: validator,
-		defaulter: defaulter,
-		tracker:   tracker,
+		client:                  client,
+		log:                     log,
+		validator:               validator,
+		defaulter:               defaulter,
+		tracker:                 tracker,
+		buildProviderReconciler: buildProviderReconciler,
 	}
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	childObjectHandler := handlerutil.ChildObjectToClusters(r.log)
```

Comment:
What's the main reason to set up ownerref? Is it mainly for GC and object adoption?
Also reason to handle enqueue request func ourselves? So the child object changes can be sent to the corresponding cluster reconciler?

Reply:
About why setting up owner ref, good question. Copy paste from the design ticket since I can't link it:


> Objects like provider machine configs, flux config, datacenter config or identity provider configs are linked in the cluster but there is no back reference.
>
> This is useful to trigger cluster reconciliation loops when this "child" objects change and to (potentially, tbd if we want this or not) delete them if the cluster is deleted and there are no more clusters referencing that object.
>
> We should explore owner references for this. It's probably the right k8s pattern, it has built it support for deletion (we can avoid an object being deleted until it's parent object is deleted) and supports many to one relationships.
>
> # Solution
>
>At the beginning of the reconciliation loop, the cluster controller will retrieve all linked objects to the `Cluster`. If it hasn't been added yet, it will add the `Cluster` as a `metadata.OwnerReferences`.
>
> ```go
>ref := metav1.OwnerReference{
>	APIVersion:         anywherev1.GroupVersion().String(),
>	Kind:               "Cluster",
>	Name:               cluster.GetName(),
>       UID:                cluster.GetUID(),
>       BlockOwnerDeletion: pointer.BoolPtr(true),
>       Controller:         pointer.BoolPtr(true),
>}
>
> owners := object.GetOwnerReferences()
> owners = append(owners, ref)
> object.SetOwnerReferences(owners)
> ```
>
> Note: cross namespace owner references are no allowed, so all eksa objects linked to a `Cluster` will need to be in the same namespace as that `Cluster`.
This is already a validation we currently run and the present code assumes this is always true. But it's worth highlighting this since we won't be able to use owner references if we decide to support cross namespace references from the `Cluster`.

What do you mean exactly by `Also reason to handle enqueue request func ourselves?`?
We are just determining which clusters need to be reconciled when a child object changes. Were you thinking about another alternative?